be/src/storage/lake/vacuum.cpp

```diff
@@ -631,4 +633,232 @@ void delete_txn_log(TabletManager* tablet_mgr, const DeleteTxnLogRequest& reques
     delete_files_async(files_to_delete);
 }
 
+static Status list_tablet_metadata(const std::string& metadir, std::set<std::string>* metas) {
+    auto fs_or = FileSystem::CreateSharedFromString(metadir);
+    if (!fs_or.ok()) {
+        // Return NotFound only when the file or directory does not exist.
+        if (fs_or.status().is_not_found()) {
+            return Status::InternalError(fs_or.status().message());
+        } else {
+            return fs_or.status();
+        }
+    }
+
+    return (*fs_or)->iterate_dir(metadir, [&](std::string_view name) {
+        if (is_tablet_metadata(name)) {
+            metas->emplace(name);
+        }
+        return true;
+    });
+}
+
+static StatusOr<TabletMetadataPtr> get_tablet_metadata(const string& metadata_location, bool fill_cache) {
+    auto metadata = std::make_shared<TabletMetadataPB>();
+    ProtobufFile file(metadata_location);
+    RETURN_IF_ERROR(file.load(metadata.get(), fill_cache));
+    return metadata;
+}
+
+static StatusOr<TxnLogPtr> get_txn_log(const std::string& txn_log_path, bool fill_cache) {
+    auto txn_log = std::make_shared<TxnLog>();
+    ProtobufFile file(txn_log_path);
+    RETURN_IF_ERROR(file.load(txn_log.get(), fill_cache));
+    return txn_log;
+}
+
+static StatusOr<std::map<std::string, DirEntry>> find_orphan_datafiles(std::string_view root_location,
+                                                                       const std::set<std::string>& tablet_metadatas,
+                                                                       const std::vector<std::string>& txn_logs,
+                                                                       int64_t expire_seconds) {
+    ASSIGN_OR_RETURN(auto fs, FileSystem::CreateSharedFromString(root_location));
+    const auto now = std::time(nullptr);
+    const auto metadata_root_location = join_path(root_location, kMetadataDirectoryName);
+    const auto txn_log_root_location = join_path(root_location, kTxnLogDirectoryName);
+    const auto segment_root_location = join_path(root_location, kSegmentDirectoryName);
+
+    LOG(INFO) << "Start to list " << segment_root_location;
+
+    std::map<std::string, DirEntry> datafiles;
+
+    int64_t total_files = 0;
+    int64_t total_bytes = 0;
+    // List segment
+    auto st = ignore_not_found(fs->iterate_dir2(segment_root_location, [&](DirEntry entry) {
+        total_files++;
+        total_bytes += entry.size.value_or(0);
+        if (!is_segment(entry.name) && !is_del(entry.name) && !is_delvec(entry.name)) {
+            LOG(WARNING) << "Unrecognized data file " << entry.name;
+            return true;
+        }
+        if (entry.mtime.has_value()) {
+            if (now >= entry.mtime.value() + expire_seconds) {
```

Comment:
I think it is better to get `now` here instead of head of `find_orphan_datafiles`, if there is a case that `iterate_dir2` cost too much time, and `now` will be too old to used.

Reply:
entry.mtime is returned by iterate_dir2, so get now should not after iterate_dir2 ?